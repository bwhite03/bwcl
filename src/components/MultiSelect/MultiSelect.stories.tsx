import React from "react";
import { Story, Meta } from "@storybook/react";
import MultiSelect, { MultiSelectProps } from "./MultiSelect";

export default {
  title: "Forms/MultiSelect",
  component: MultiSelect,
  parameters: {
    layout: "centered",
  },
} as Meta;

let checkedItems: string[] = [];

const data = [
  {
    id: 1,
    value: "Grocery",
  },
  {
    id: 2,
    value: "Nonfood",
  },
  {
    id: 3,
    value: "HBA",
  },
  {
    id: 4,
    value: "Meat",
  },
  {
    id: 5,
    value: "Meat by #",
  },
  {
    id: 6,
    value: "Produce",
  },
  {
    id: 7,
    value: "Produce by #",
  },
];

function setCheckedItems(s: string[]) {
  checkedItems = s;
}

const Template =
  <T extends {}>(): Story<MultiSelectProps<T>> =>
  (args) =>
    <MultiSelect {...args} />;

export const MultiSelectExample = Template<object>().bind({});
MultiSelectExample.args = {
  data,
  defaultOption: "Select a department",
  keyDescriptior: "ms-cb-1",
  // @ts-ignore
  optionValue: "id",
  // @ts-ignore
  optionText: "value",
  checkedItems: checkedItems,
  setCheckedItems: setCheckedItems,
};
