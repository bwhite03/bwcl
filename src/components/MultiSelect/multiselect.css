.select-box {
  position: relative;
}

.multi-select-checkboxes {
  position: absolute;
  top: 0px;
  left: 0px;
  border: 1px solid #2e3332;
  z-index: 999;
  max-height: 300px;
  overflow-y: scroll;
  animation: fadeInMultiSelectPortal 1s ease;
}

.multi-select-checkboxes label {
  display: block;
  padding: 3px 8px;
  color: #fff;
}

.multi-select-checkboxes input {
  margin-right: 5px;
}

.multi-select-checkboxes input:focus {
  outline: none;
  border: 0;
}

@keyframes fadeInMultiSelectPortal {
  0% {
    opacity: 0;
  }

  100% {
    opacity: 1;
  }
}
