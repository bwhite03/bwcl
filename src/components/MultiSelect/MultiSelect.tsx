import React, { useState, useEffect } from 'react';
import MultiSelectPortal from '../MultiSelectPortal/MultiSelectPortal';
import './multiselect.css';

export interface MultiSelectProps<T> {
  data: T[];
  defaultOption: string;
  keyDescriptior: string;
  optionValue: keyof T;
  optionText: keyof T;
  checkedItems: string[];
  setCheckedItems: (e: string[]) => void;
  id?: string;
  selectBoxId?: string;
  backgroundColor?: string;
  selectId?: string;
  inputId?: string;
}

function MultiSelect<T>(props: MultiSelectProps<T>) {
  const [show, setShow] = useState(false);
  const [filteredData, setFilteredData] = useState<T[]>([]);
  const [search, setSearch] = useState('');
  const [locallyCheckedItems, setLocallyCheckedItems] = useState<string[]>([]);

  const {
    id = 'multi-select-checkboxes',
    selectBoxId = 'select-box-id',
    backgroundColor = '#1f1e2e',
    selectId = 'select-id',
    inputId = 'multi-select-search',
    checkedItems,
    optionValue,
    keyDescriptior,
    optionText,
    setCheckedItems,
  } = props;

  useEffect(() => {
    setFilteredData(props.data);
    setLocallyCheckedItems(checkedItems);
  }, [props.data]);

  const amIChecked = (i: T) => {
    if (locallyCheckedItems.includes(i[optionValue] as unknown as string)) {
      return true;
    } else {
      return false;
    }
  };

  const showCheckBoxes = () => {
    if (show) {
      setShow(false);
      return;
    } else {
      setShow(true);
    }

    const actualSelect = document.getElementById(selectId) as HTMLSelectElement;
    if (actualSelect) {
      actualSelect.options.remove;

      setTimeout(() => {
        const select = document.getElementById(selectBoxId);
        const portal = document.getElementById(id);
        if (select && portal) {
          const portalRect = portal.getBoundingClientRect();
          const selectRect = select.getBoundingClientRect();
          portal.style.top = selectRect.bottom + 'px';
          portal.style.left = selectRect.left + 'px';
          portal.style.backgroundColor = backgroundColor;
          window.addEventListener('click', tapOutside);
          const input = document.getElementById(inputId);
          if (input) {
            input.focus();
          }
        }
      }, 100);
    }
  };

  const tapOutside = (e: MouseEvent) => {
    const portal = document.getElementById(id);
    if (portal) {
      const rect = portal?.getBoundingClientRect();
      if (e.clientX < rect.left || e.clientX > rect.right) {
        setShow(false);
        //   const actualSelect = document.getElementById(selectId);
        //   const option = new Option(props.defaultOption, "0");
        //   actualSelect?.prepend(option);
        window.removeEventListener('click', tapOutside);
        return;
      }
      if (e.clientY < rect.top || e.clientY > rect.bottom) {
        setShow(false);
        //   const actualSelect = document.getElementById(selectId);
        //   const option = new Option(props.defaultOption, "0");
        //   actualSelect?.prepend(option);
        window.removeEventListener('click', tapOutside);
        return;
      } else {
        window.removeEventListener('click', tapOutside);
      }
    }
  };

  const handleCheck = (e: React.ChangeEvent<HTMLInputElement>, i: T) => {
    const columnValue = i[optionValue] as unknown as string;
    if (e.target.checked) {
      if (!locallyCheckedItems.includes(columnValue)) {
        setCheckedItems([...checkedItems, columnValue]);
        setLocallyCheckedItems([...locallyCheckedItems, columnValue]);
      }
    } else {
      const newArray = locallyCheckedItems.filter((i) => i !== columnValue);
      setCheckedItems(newArray);
      setLocallyCheckedItems(newArray);
    }
  };

  const handleSearchChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    setSearch(e.target.value);
    if (e.target.value.length === 0) {
      setFilteredData(props.data);
    } else {
      const filtered = props.data.filter((s) => {
        const recordValue = s[optionText] as unknown as string;
        if (
          recordValue.toLowerCase().startsWith(e.target.value.toLowerCase())
        ) {
          return s;
        }
      });
      setFilteredData(filtered);
    }
  };

  return (
    <div className="multi-select">
      <div
        className="select-box"
        id={selectBoxId}
        onClick={() => showCheckBoxes()}
      >
        <select className="form-control" id={selectId}>
          <option value="0">{props.defaultOption}</option>
        </select>
      </div>
      <div id="ms-portal-container"></div>
      <MultiSelectPortal
        open={show}
        data={filteredData}
        id={id}
        keyDescriptor={keyDescriptior}
        optionText={optionText}
        handleCheck={handleCheck}
        handleSearchChange={handleSearchChange}
        search={search}
        optionValue={optionValue}
        amIChecked={amIChecked}
        inputId={inputId}
      />
    </div>
  );
}

export default MultiSelect;
