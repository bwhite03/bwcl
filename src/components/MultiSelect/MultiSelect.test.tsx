import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import MultiSelect from './MultiSelect';

const data = [
  {
    id: 1,
    value: 'Grocery',
  },
  {
    id: 2,
    value: 'Nonfood',
  },
  {
    id: 3,
    value: 'HBA',
  },
  {
    id: 4,
    value: 'Meat',
  },
  {
    id: 5,
    value: 'Meat by #',
  },
  {
    id: 6,
    value: 'Produce',
  },
  {
    id: 7,
    value: 'Produce by #',
  },
];

describe('MultiSelect', () => {
  test('should render correctly', () => {
    const setCheckedItems = jest.fn();
    jest.useFakeTimers();
    jest.spyOn(global, 'setTimeout');
    const { container } = render(
      <MultiSelect
        data={data}
        defaultOption="select a department"
        keyDescriptior="keyDescriptor"
        optionValue="id"
        optionText="value"
        checkedItems={[]}
        setCheckedItems={setCheckedItems}
        id="id"
        selectBoxId="selectBoxId"
        backgroundColor="red"
        selectId="selectId"
        inputId="inputId"
      />
    );

    expect(container).toMatchSnapshot();
    expect(setTimeout).toHaveBeenCalled();
  });

  test('should rener without default props', () => {
    const setCheckedItems = jest.fn();
    const { container, getByRole } = render(
      <MultiSelect
        data={data}
        defaultOption="select a department"
        keyDescriptior="keyDescriptor"
        optionValue="id"
        optionText="value"
        checkedItems={[]}
        setCheckedItems={setCheckedItems}
        id="id"
        selectBoxId="selectBoxId"
        backgroundColor="red"
        selectId="selectId"
        inputId="inputId"
      />
    );

    expect(container).toMatchSnapshot();
    const option = getByRole('option', { name: /select a department/ });
    fireEvent.click(option);
    fireEvent.click(option);
  });

  test('should handle click outside', () => {
    const div = document.createElement('div');
    div.id = 'outside';
    document.body.appendChild(div);
    const setCheckedItems = jest.fn();
    const { container, getByRole } = render(
      <MultiSelect
        data={data}
        defaultOption="select a department"
        keyDescriptior="keyDescriptor"
        optionValue="id"
        optionText="value"
        checkedItems={[]}
        setCheckedItems={setCheckedItems}
        id="id"
        selectBoxId="selectBoxId"
        backgroundColor="red"
        selectId="selectId"
        inputId="inputId"
      />
    );

    const option = getByRole('option', { name: /select a department/ });
    fireEvent.click(option);
    const outside = document.querySelector('outside');
    if (outside) {
      fireEvent.click(outside);
    }
  });

  test('should handle check', () => {
    const div = document.createElement('div');
    div.id = 'ms-portal-container';
    document.body.appendChild(div);
    const setCheckedItems = jest.fn();
    const { container, getByRole } = render(
      <MultiSelect
        data={data}
        defaultOption="select a department"
        keyDescriptior="keyDescriptor"
        optionValue="id"
        optionText="value"
        checkedItems={[]}
        setCheckedItems={setCheckedItems}
        id="id"
        selectBoxId="selectBoxId"
        backgroundColor="red"
        selectId="selectId"
        inputId="inputId"
      />
    );

    const option = getByRole('option', { name: /select a department/ });
    fireEvent.click(option);

    const dept = getByRole('checkbox', { name: /Grocery/ });
    fireEvent.click(dept);
    fireEvent.click(dept);
  });

  test('should handle search', () => {
    const div = document.createElement('div');
    div.id = 'ms-portal-container';
    document.body.appendChild(div);
    const setCheckedItems = jest.fn();
    const { container, getByRole, getByPlaceholderText } = render(
      <MultiSelect
        data={data}
        defaultOption="select a department"
        keyDescriptior="keyDescriptor"
        optionValue="id"
        optionText="value"
        checkedItems={[]}
        setCheckedItems={setCheckedItems}
        id="id"
        selectBoxId="selectBoxId"
        backgroundColor="red"
        selectId="selectId"
        inputId="inputId"
      />
    );

    const option = getByRole('option', { name: /select a department/ });
    fireEvent.click(option);

    userEvent.type(screen.getByPlaceholderText(/search/i), 'gro');
    const input = getByPlaceholderText('search') as HTMLInputElement;
    expect(input.value).toEqual('gro');

    userEvent.type(
      screen.getByPlaceholderText(/search/i),
      '{backspace}{backspace}{backspace}'
    );

    expect(input.value).toEqual('');
  });
});
