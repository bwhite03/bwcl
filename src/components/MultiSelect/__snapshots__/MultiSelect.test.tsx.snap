// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiSelect should render correctly 1`] = `
<div>
  <div
    class="multi-select"
  >
    <div
      class="select-box"
      id="selectBoxId"
    >
      <select
        class="form-control"
        id="selectId"
      >
        <option
          value="0"
        >
          select a department
        </option>
      </select>
    </div>
    <div
      id="ms-portal-container"
    />
  </div>
</div>
`;

exports[`MultiSelect should rener without default props 1`] = `
<div>
  <div
    class="multi-select"
  >
    <div
      class="select-box"
      id="selectBoxId"
    >
      <select
        class="form-control"
        id="selectId"
      >
        <option
          value="0"
        >
          select a department
        </option>
      </select>
    </div>
    <div
      id="ms-portal-container"
    />
  </div>
</div>
`;
